bin/elasticurl/main.cpp

```diff
@@ -0,0 +1,540 @@
+    if (input_body == nullptr)
+    {
+        input_body = std::make_shared<std::stringstream>("");
+    }
+
+    if (aws_cli_optind < argc)
+    {
+        struct aws_byte_cursor uri_cursor = aws_byte_cursor_from_c_str(argv[aws_cli_optind++]);
+
+        if (aws_uri_init_parse(&ctx->uri, ctx->allocator, &uri_cursor))
+        {
+            fprintf(
+                stderr,
+                "Failed to parse uri %s with error %s\n",
+                (char *)uri_cursor.ptr,
+                aws_error_debug_str(aws_last_error()));
+            s_usage(1);
+        };
+    }
+    else
+    {
+        fprintf(stderr, "A URI for the request must be supplied.\n");
+        s_usage(1);
+    }
+}
+
+int main(int argc, char **argv)
+{
+    struct aws_allocator *allocator = aws_default_allocator();
+    (void)allocator;
+    (void)argc;
+    (void)argv;
+
+    struct elasticurl_ctx app_ctx;
+    AWS_ZERO_STRUCT(app_ctx);
+    app_ctx.allocator = allocator;
+    app_ctx.connect_timeout = 3000;
+    app_ctx.output = stdout;
+    app_ctx.verb = "GET";
+    app_ctx.alpn = "h2;http/1.1";
+
+    s_parse_options(argc, argv, &app_ctx);
+
+    Aws::Crt::ApiHandle apiHandle(allocator);
+    if (app_ctx.trace_file)
+    {
+        apiHandle.InitializeLogging(app_ctx.log_level, app_ctx.trace_file);
+    }
+    else
+    {
+        apiHandle.InitializeLogging(app_ctx.log_level, stdout);
+    }
+    bool use_tls = true;
+    uint16_t port = 443;
+    if (!app_ctx.uri.scheme.len && (app_ctx.uri.port == 80 || app_ctx.uri.port == 8080))
+    {
+        use_tls = false;
+    }
+    else
+    {
+        if (aws_byte_cursor_eq_c_str_ignore_case(&app_ctx.uri.scheme, "http"))
+        {
+            use_tls = false;
+        }
+    }
+
+    auto hostName = app_ctx.uri.host_name;
+
+    Aws::Crt::Io::TlsContextOptions tlsCtxOptions;
+    Aws::Crt::Io::TlsContext tlsContext;
+    Aws::Crt::Io::TlsConnectionOptions tlsConnectionOptions;
+    if (use_tls)
+    {
+        if (app_ctx.cert && app_ctx.key)
+        {
+            tlsCtxOptions = Aws::Crt::Io::TlsContextOptions::InitClientWithMtls(app_ctx.cert, app_ctx.key);
+        }
+        else
+        {
+            tlsCtxOptions = Aws::Crt::Io::TlsContextOptions::InitDefaultClient();
+        }
+        if (app_ctx.capath || app_ctx.cacert)
+        {
+            tlsCtxOptions.OverrideDefaultTrustStore(app_ctx.capath, app_ctx.cacert);
+        }
+        if (app_ctx.insecure)
+        {
+            tlsCtxOptions.SetVerifyPeer(false);
+        }
+
+        tlsContext = Aws::Crt::Io::TlsContext(tlsCtxOptions, Aws::Crt::Io::TlsMode::CLIENT, allocator);
+
+        tlsConnectionOptions = tlsContext.NewConnectionOptions();
+
+        tlsConnectionOptions.SetServerName(hostName);
+        tlsConnectionOptions.SetAlpnList(app_ctx.alpn);
+    }
+    else
+    {
+        if (app_ctx.required_http_version == AWS_HTTP_VERSION_2)
+        {
+            fprintf(stderr, "Error, we don't support h2c, please use TLS for HTTP2 connection");
```

Comment:
should all those fprintf be converted to c++ style? std::cerr<<?

Reply:
And I ignored some error handling for now, I'll fix them as I finish this up

Reply:
sure
honestly C++ programmers often use printf over std::cerr, but I think in this case std::cerr will be simpler

dont forget to end lines wih "\n" or `<< std::endl;`